presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EvaluateProjectOnValues.java

```diff
@@ -0,0 +1,144 @@
+    }
+
+    @Override
+    public Result apply(ProjectNode projectNode, Captures captures, Context context)
+    {
+        Optional<PlanNode> optionalSource = context.getLookup().resolveGroup(projectNode.getSource()).findFirst();
+        if (!optionalSource.isPresent()) {
+            return Result.empty();
+        }
+        ValuesNode source = (ValuesNode) optionalSource.get();
+        List<List<RowExpression>> rows = source.getRows();
+        List<VariableReferenceExpression> valuesOutputVariables = source.getOutputVariables();
+        Set<Map.Entry<VariableReferenceExpression, RowExpression>> projectAssignmentEntries = projectNode.getAssignments().entrySet();
+        List<VariableReferenceExpression> projectOutputVariables = projectAssignmentEntries.stream().map(Map.Entry::getKey).collect(toImmutableList());
+        List<RowExpression> projectRowExpressions = projectAssignmentEntries.stream().map(Map.Entry::getValue).collect(toImmutableList());
+        // exclude non-deterministic function
+        if (!projectRowExpressions.stream()
+                .filter(expression -> expression instanceof CallExpression)
+                .map(callExpression -> functionAndTypeManager.getFunctionMetadata(((CallExpression) callExpression).getFunctionHandle()))
+                .allMatch(FunctionMetadata::isDeterministic)) {
+            return Result.empty();
+        }
+
+        //evaluate ProjectNode
+        List<List<Object>> rowValues = rows.stream().map(row -> {
+            //Prepare to set up variable resolver
+            verify(row.size() == valuesOutputVariables.size(), "Output variable does not match its value in ValuesNode");
+            Map<String, Object> valuesMapForResolver = Streams.zip(valuesOutputVariables.stream(),
+                    row.stream(),
+                    (valuesOutputVariable, element) -> {
+                        RowExpressionInterpreter elementInterpreter = new RowExpressionInterpreter(element,
+                                functionAndTypeManager,
+                                context.getSession().toConnectorSession(),
+                                EVALUATED);
+                        return new AbstractMap.SimpleImmutableEntry<String, Object>(valuesOutputVariable.getName(), elementInterpreter.evaluate());
+                    }).collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
+            VariableResolver variableResolver = new Interpreters.LambdaVariableResolver(valuesMapForResolver);
+            //evaluate each row of the ProjectNode
+            return projectRowExpressions.stream().map(rowExpression -> new RowExpressionInterpreter(
+                            rowExpression,
+                            functionAndTypeManager,
+                            context.getSession().toConnectorSession(),
+                            OPTIMIZED).optimize(variableResolver))
+                    .collect(toImmutableList());
+        }).collect(toImmutableList());
```

Comment:
This part seems a bit complex and I have a hard time to understand it. Can you check the `ExpressionRewriter` in `CommonSubExpressionRewriter` class, and see if you can rewrite the expression in project assignments with values in Values node, and it should be later evaluated by `RowExpressionOptimizer`.

Reply:
I used _ExpressionRewriter_ to rewrite the expression in ProjectNode. However, evaluating _ProjectNode_ requires evaluating _ValuesNode_ first to provide the resolver to the _optimize_ function of _RowExpressionOptimizer_. The overall implementation does not simplify too much.

Reply:
If I'm understanding correctly, I think what we actually want here is the `RowExpressionVariableInliner`? The `ExpressionRewriter` replaces a `RowExpression` with a `VariableReferenceExpression`. I think we want the other way around when constructing expressions to evaluate. If we can re-write the projection input expression to replace the reference to the reference coming from the ValuesNode, then we only need to run the Interpreter once. Is that correct @feilong-liu ?

Reply:
@ZacBlanco You are right, we need to use RowExpressionVariableInliner here.

@jackychen718  Can you simplify the logic here with the RowExpressionVariableInliner?

Also is it possible to get rid of the RowExpressionInterpreter in this optimizer? I think the RowExpressionOptimizer which ran later should be able to evaluate it (worth to try e2e to verify)

Reply:
fixed